sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/data/BeamFnDataSizeBasedBufferingOutboundObserver.java

```diff
@@ -85,7 +85,7 @@ public void close() throws Exception {
   }
 
   @Override
-  public void flush() throws IOException {
+  public synchronized void flush() throws IOException {
```

Comment:
We don't want to make this synchronized since this class is not thread safe and should not take this perf hit. Your previous usage of defining an override that is synchronized for the class that called super.flush() is all you need.